From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NotCoded <66999075+not-coded@users.noreply.github.com>
Date: Mon, 18 May 2026 21:58:49 +0200
Subject: [PATCH] Map Combat Test 8c Bukkit API


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index e2b3470e3c9a97671723f5a67f722fb86fb07fbf..c0dc4f30bb2d4f9921622df79fefa035878eed9a 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -1779,7 +1779,7 @@ public enum Material implements Keyed {
     MINECART(14352, 1),
     SADDLE(30206, 1),
     REDSTONE(11233),
-    SNOWBALL(19487, 16),
+    SNOWBALL(19487, 64),
     OAK_BOAT(17570, 1),
     LEATHER(16414),
     MILK_BUCKET(9680, 1),
@@ -1916,7 +1916,7 @@ public enum Material implements Keyed {
      * BlockData: {@link Ageable}
      */
     NETHER_WART(29227, Ageable.class),
-    POTION(24020, 1),
+    POTION(24020, 16),
     GLASS_BOTTLE(6116),
     SPIDER_EYE(9318),
     FERMENTED_SPIDER_EYE(19386),
diff --git a/src/main/java/org/bukkit/attribute/Attribute.java b/src/main/java/org/bukkit/attribute/Attribute.java
index 13eac9ad2c1672051635d1c35cc49239252e7a61..cfd927edc3fc7f824f3b27a28e46069a950e8e8f 100644
--- a/src/main/java/org/bukkit/attribute/Attribute.java
+++ b/src/main/java/org/bukkit/attribute/Attribute.java
@@ -41,6 +41,10 @@ public enum Attribute implements Keyed {
      * Attack speed of an Entity.
      */
     GENERIC_ATTACK_SPEED("generic.attack_speed"),
+    /**
+     * Attack reach of an Entity.
+     */
+    GENERIC_ATTACK_REACH("generic.attack_reach"),
     /**
      * Armor bonus of an Entity.
      */
diff --git a/src/main/java/org/bukkit/enchantments/Enchantment.java b/src/main/java/org/bukkit/enchantments/Enchantment.java
index 799813410f40589ba1d7d530a47d87f609764705..3bdc0feda1642e1681506493cbbc5dc48c473568 100644
--- a/src/main/java/org/bukkit/enchantments/Enchantment.java
+++ b/src/main/java/org/bukkit/enchantments/Enchantment.java
@@ -103,6 +103,11 @@ public abstract class Enchantment implements Keyed {
      */
     public static final Enchantment SWEEPING_EDGE = new EnchantmentWrapper("sweeping");

+    /**
+     * Increases axe melee damage and shield disable time.
+     */
+    public static final Enchantment CLEAVING = new EnchantmentWrapper("cleaving");
+
     /**
      * Increases the rate at which you mine/dig
      */
diff --git a/src/main/java/org/bukkit/enchantments/EnchantmentTarget.java b/src/main/java/org/bukkit/enchantments/EnchantmentTarget.java
index 635e07a6b0e255c4fdad58ba9d281c807af4e229..e52807afeda6a4bc1eb8bec457e6200ca7804f57 100644
--- a/src/main/java/org/bukkit/enchantments/EnchantmentTarget.java
+++ b/src/main/java/org/bukkit/enchantments/EnchantmentTarget.java
@@ -116,6 +116,21 @@ public enum EnchantmentTarget {
         }
     },

+    /**
+     * Allows the Enchantment to be placed on axes.
+     */
+    AXE {
+        @Override
+        public boolean includes(@NotNull Material item) {
+            return item.equals(Material.WOODEN_AXE)
+                || item.equals(Material.STONE_AXE)
+                || item.equals(Material.IRON_AXE)
+                || item.equals(Material.DIAMOND_AXE)
+                || item.equals(Material.GOLDEN_AXE)
+                || item.equals(Material.NETHERITE_AXE);
+        }
+    },
+
     /**
      * Allows the Enchantment to be placed on tools (spades, pickaxe, axes)
      */
diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -306,5 +306,6 @@ public interface HumanEntity extends InventoryHolder, AnimalTamer, Permissible {
-     * This is used to calculate damage, with 1.0 representing a fully charged
-     * attack and 0.0 representing a non-charged attack
+     * This is used to calculate damage, with 0.0 representing a non-charged
+     * attack, 1.0 representing a fully charged attack, and 2.0 representing a
+     * charge of 200%.
      *
-     * @return A float between 0.0-1.0 representing the progress of the charge
+     * @return A float between 0.0-2.0 representing the progress of the charge
      */
